### PR TITLE
Add helpers for Amazon, Intel and Android OS detection

### PR DESCRIPTION
**Pull request by zegnus**
## Feature description

Android applications can run in a variety of different providers and manufacturers and sometimes we need to provide different set of supported features depending on the user's device.

This Pull Requests adds a couple of helper classes:

**DeviceDetection**
Will provide you help for evaluating if the user's device is an Intel or an Amazon powered device
Usage:
```java
DeviceDetection deviceDetection = DeviceDetection.newInstance();
boolean intel = deviceDetection.is(DeviceDetection.INTEL); // for Intel powered devices
boolean = amazon = deviceDetection.is(DeviceDetection.AMAZON); // for Amazon manufactured devices
```

As a helper of the DeviceDetection, another class has been provided:

**AndroidVersion**
Will provide you direct methods for querying the API level of the current device
Usage:
```java
AndroidVersion androidVersion = AndroidVersion.newInstance();
boolean is18JellyBeanOrOver = androidVersion.is18JellyBeanOrOver();
boolean is19KitKat = androidVersion.is19KitKat();
boolean is19KitKatOrOver = androidVersion.is19KitKatOrOver();
boolean is21LollipopOrOver = androidVersion.is21LollipopOrOver();
boolean isPre21Lollipop = androidVersion.isPre21Lollipop();
boolean is23MarshmallowOrOver = androidVersion.is23MarshmallowOrOver();

int apiLevel = androidVersion.apiLevel(); // will return Build.VERSION.SDK_INT
int versionName = androidVersion.versionName(); // will return Build.VERSION.RELEASE
```

### CODE REVIEW

File: android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.novoda.notils.devicedetection;
+
+class AmazonDetector {
+
+    //From: https://developer.amazon.com/appsandservices/solutions/devices/kindle-fire/specifications/01-device-and-feature-specifications
+    private static final String AMAZON = "Amazon";
+
+    private final String buildManufacturer;
+
+    AmazonDetector(String buildManufacturer) {
```
Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java]: would prefer if the `buildManufacturer` was the method param instead of ctor param. There's two cases this could be used:

- checking the device you're running on
- checking a list of buildManufacturers you get

For the first case, either way works, but for the second case, passing the `buildManufacturer` into the ctor means instantiating a new one each time you want to check.

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,6 +1,6 @@
-#Wed May 21 16:59:51 BST 2014
+#Tue Aug 09 11:54:01 CEST 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=http\://services.gradle.org/distributions/gradle-2.2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.2-all.zip
```
Comment (rock3r) [inline gradle/wrapper/gradle-wrapper.properties]: We could probably update this :P

File: android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package com.novoda.notils.devicedetection;
+
+import android.os.Build;
+
+public class DeviceAndroidVersion {
```
Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java]: `AndroidVersion`? it's not tied to the device. Same for the field, and then syncing up the class name and the field name (including `os` or omitting it)

Comment (rock3r) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java]: This seems not to be the most recent version we have internally. ⚠️

File: android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java (hunk 1 above)
Comment (zegnus) [inline android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java]: Consider that this class is package protected and not part of the public API.

For the second scenario one way to do it would be to pass an array of manufacturers instead of just one value ;)

Also this is in sync with the `X86Detector` that receives a bunch of values at constructor time that are calculated based on the platform. And we only want to do that once and not every time that we check.

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (zegnus) [inline gradle/wrapper/gradle-wrapper.properties]: let me do this in another PR

File: android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java (hunk 3 above)
Comment (zegnus) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java]: f2e35e94fe9320fd2deff3677cf5c480b6af9547

Comment (zegnus) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceAndroidVersion.java]: 13d8200bfe53d613f05d08d0e4edc2767a08206c

File: android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java (hunk 1 above)
Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java]: >Consider that this class is package protected and not part of the public API.

Yep but since it's going into notils, which is a collection of utility classes, we might not want to couple the classes together. Consider the case where I don't care about x86, and just want an Amazon device detector - the Amazon detector doesn't rely on anything, but I'm forced to create the DeviceDetector, and worse, I'm unable to achieve the second scenario above because I have no way to pass anything in.

>For the second scenario one way to do it would be to pass an array of manufacturers instead of just one value ;)
>
>Also this is in sync with the X86Detector that receives a bunch of values at constructor time that are calculated based on the platform. And we only want to do that once and not every time that we check.

Yep would also argue the same for the x86Detector, the array of supported ABIs should be a method param. The DeviceDetector can still create that list once and not every time we check, but we move that requirement to the deviceDetector, not into these classes.

File: android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java [hunk 4]
```diff
@@ -1,18 +1,40 @@
...
 import java.util.Arrays;
 
-class X86Detector {
+class X86Detector implements Detector {
 
     private static final String X86 = "x86";
 
     private final String[] supportedAbis;
 
+    public static Detector newInstance() {
```
Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java]: :+1: I think this is clearer, keeping this internal.

Not sure the interface adds anything though.

File: android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.novoda.notils.devicedetection;
+
+import android.os.Build;
+
+class AmazonDetector implements Detector {
+
+    //From: https://developer.amazon.com/appsandservices/solutions/devices/kindle-fire/specifications/01-device-and-feature-specifications
```
Comment (blundell) [inline android/src/main/java/com/novoda/notils/devicedetection/AmazonDetector.java]: should use a proper javadoc comment

```
/**
*
**/
```

File: android/src/main/java/com/novoda/notils/devicedetection/DeviceDetection.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package com.novoda.notils.devicedetection;
+
+import android.support.annotation.IntDef;
+import android.util.SparseArray;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+public class DeviceDetection {
```
Comment (blundell) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceDetection.java]: should have javadoc since this is the public facing api

File: android/src/main/java/com/novoda/notils/devicedetection/DeviceDetection.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static final int INTEL = 0;
```
Comment (blundell) [inline android/src/main/java/com/novoda/notils/devicedetection/DeviceDetection.java]: why are these public? should be private

File: android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    private static String[] getSupportedAbis() {
+        return Build.SUPPORTED_ABIS;
+    }
+
+    private static String[] getLegacySupportedAbis() {
+        return new String[]{Build.CPU_ABI, Build.CPU_ABI2};
+    }
+
+    X86Detector(String... supportedAbis) {
+        this.supportedAbis = Arrays.copyOf(supportedAbis, supportedAbis.length);
```
Comment (blundell) [inline android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java]: why make the copy here? Makes the class harder to test - should be done in the newInstance method so that the constr can be used easily for testing

Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java]: PMS/findbugs will flag it here again if you move the copy to the ctor (not sure if it will flag it if the ctor is private). This ctor can still be used for testing - the values will still be same, you're not gonna mock a string array.

Comment (blundell) [inline android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java]: good points sir! 💯

Comment (ataulm) [inline android/src/main/java/com/novoda/notils/devicedetection/X86Detector.java]: :tada: and autocorrect I meant PMD

File: android/src/main/java/com/novoda/notils/devicedetection/AndroidVersion.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    public boolean is21LollipopOrOver() {
+        return osApiLevel >= Build.VERSION_CODES.LOLLIPOP;
+    }
+
+    public boolean isPre21Lollipop() {
+        return !is21LollipopOrOver();
+    }
+
+    public boolean is23MarshmallowOrOver() {
+        return osApiLevel >= Build.VERSION_CODES.M;
```
Comment (tasomaniac) [inline android/src/main/java/com/novoda/notils/devicedetection/AndroidVersion.java]: When you have `if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)` in code, android lint detects it and does not warn api level related stuff that requires `M` inside that if statement. 

But if you wrap it like this, unfortunately that link check doesn't detect and you will end-up using `@TargetApi` annotation to suppress the warning. 

`@TargetApi` means that the method/class is targeted only to that api level. But in most cases they are not. They are used for all devices but with slight modification on some api levels.
